robot/impl/local_robot.go

```diff
@@ -1190,42 +1193,13 @@ func (r *localRobot) applyLocalModuleVersions(cfg *config.Config) {
 }
 
 func (r *localRobot) reconfigure(ctx context.Context, newConfig *config.Config, forceSync bool) {
-	// Maintenance config can be configured to block reconfigure based off of a sensor reading
-	// These sensors can be configured on the main robot, or a remote
-	// In situations where there are conflicting sensor names the following behavior happens
-	// Main robot and remote share sensor name -> main robot sensor is chosen
-	// Only remote has the sensor name -> remote sensor is read
-	// Multiple remotes share a senor name -> conflict error is returned and reconfigure happens
-	// To specify a specific remote sensor use the name format remoteName:sensorName to specify a remote sensor
-	if newConfig.MaintenanceConfig != nil {
-		name, err := resource.NewFromString(newConfig.MaintenanceConfig.SensorName)
-		if err != nil {
-			r.logger.Warnf("sensor_name %s in maintenance config is not in a supported format", newConfig.MaintenanceConfig.SensorName)
-		} else {
-			sensorComponent, err := robot.ResourceFromRobot[sensor.Sensor](r, name)
-			if err != nil {
-				r.logger.Warnf("%s, Starting reconfiguration", err.Error())
-			} else {
-				canReconfigure, err := r.checkMaintenanceSensorReadings(ctx, newConfig.MaintenanceConfig.MaintenanceAllowedKey, sensorComponent)
-				if !canReconfigure {
-					if err != nil {
-						r.logger.CErrorw(ctx, "error reading maintenance sensor", "error", err)
-					} else {
-						r.logger.Info("maintenance_allowed_key found from readings on maintenance sensor. Skipping reconfiguration.")
-					}
-					diff, err := config.DiffConfigs(*r.Config(), *newConfig, false)
-					if err != nil {
-						r.logger.CErrorw(ctx, "error diffing the configs", "error", err)
-					}
-					// NetworkEqual checks if Cloud/Auth/Network are equal between configs
-					if diff != nil && !diff.NetworkEqual {
-						r.logger.Info("Machine reconfiguration skipped but Cloud/Auth/Network config section contain changes and will be applied.")
-					}
-					return
-				}
-				r.logger.Info("maintenance_allowed_key found from readings on maintenance sensor. Starting reconfiguration")
-			}
-		}
+	r.reconfiguring.Store(true)
+	defer func() {
+		r.reconfiguring.Store(false)
+	}()
+
+	if !r.reconfigureAllowed(ctx, newConfig, true) {
```

Comment:
> but I probably would have opted for doing "reconfigure allowed" first.

Good point and catch here. Apologies for the wordiness below; TL;DR I think `reconfiguring` should be `reconfiguringMu` and we need some extra logic in the agent (forthcoming PR) to be extra defensive here.

I have opted to make `r.reconfiguring` a mutex: `r.reconfiguringMu`. I do not think the `r.reconfiguring` atomic boolean is "enough" to stop `RestartAllowed` from returning true when the local robot is reconfiguring, even if I were to switch the order to what you are suggesting. Let's say this order of events occurs:

1. `RestartAllowed` checks `r.reconfiguring` and finds it to be false
2. `reconfigure` sets `r.reconfiguring` to be true
3. `RestartAllowed` checks the maintenance sensor and finds it to allow reconfiguration
4. `reconfigure` checks the maintenance sensor and finds it allow reconfiguration
5. `reconfigure` begins
6. `RestartAllowed` returns true (external restart is allowed)
7. `viam-agent` restarts `viam-server` in the middle of a reconfiguration

This will come up again the `viam-agent` PR that checks the `restart_status` value, but I think, even with a mutex, there will still be a race between `restart_status` indicating the `viam-server` is allowing an external restart and `viam-server` beginning a reconfiguration shortly after that `restart_status` response. Per the scope, we can at least alleviate that issue by checking that `restart_status` indicates the `viam-server` is allowing an external restart **four times over the course of a minute**. The thought there being that if a `viam-server` upgrade _and_ reconfiguration happen as part of a single updated config, the `viam-server` would begin its reconfiguration within the course of a minute, and one of the four `restart_status` checks would indicate that the `viam-server` is _not_ allowing an external restart.